### PR TITLE
$range language extension

### PR DESCRIPTION
**Pull request by tomblind**
This adds the `$range` language extension as a replacement for the `@forRange` annotation, as mentioned in #963

### CODE REVIEW

File: src/transformation/utils/diagnostics.ts [hunk 1]
```diff
@@ -86,6 +86,14 @@ export const invalidForRangeCall = createErrorDiagnosticFactory(
     (message: string) => `Invalid @forRange call: ${message}.`
 );
 
+export const invalidRangeUse = createErrorDiagnosticFactory(
+    "$range can only be used as a single expression in a for...of loop."
```
Comment (Perryvw) [inline src/transformation/utils/diagnostics.ts:90]: Not sure what you mean with `as a single expression`, maybe just remove that part?

File: test/unit/language-extensions/range.spec.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		for (i of $range(1, 5)) {}
+	`
+        .setOptions(rangeProjectOptions)
+        .expectDiagnosticsToMatchSnapshot([invalidRangeControlVariable.code]);
+});
+
+test.each([
+    ["for (const i in $range(1, 10, 2)) {}", [invalidRangeUse.code]],
+    ["const x = $range(1, 10);", [invalidRangeUse.code]],
+    ["const range = $range;", [invalidRangeUse.code, invalidRangeUse.code]],
```
Comment (Perryvw) [inline test/unit/language-extensions/range.spec.ts:58]: Same code mentioned twice here, could that be reduced to 1? Then is there a need to specify them as part of the each() at all or could you just put it in the test body?

Comment (tomblind) [reply test/unit/language-extensions/range.spec.ts]: This is an odd side effect of how the $range function was being detected. I found a way to do that better and avoid the redundant diagnostic.

**Comment by lolleko** [comment]
I think this should also deprecate `@forRange`.

For that push a `annotationDeprecated ` diagnostic when it's used and add a test to https://github.com/TypeScriptToLua/TypeScriptToLua/blob/master/test/unit/annotations/deprecated.spec.ts

**Review by lolleko — APPROVED** [review]

**Pull request merged**